src/RWXLoader.js

```diff
@@ -2358,6 +2351,17 @@ class RWXLoader extends Loader {
 
 				if ( tprops.length == 16 ) {
 
+					// Avoid singular matrices by keeping scale values non-zero as they can cause problems in transformations
+					[ 1, 5, 9 ].forEach( ( index ) => {
```

Comment:
In my case, I had to change the second and sixth values, but that's a very good question.
Changing the code with `[ 0, 5, 10 ]` doesn't solve the problem, so either this transformation matrix isn't standard, or my assumption is completely wrong.

Reply:
The thing is: just fitting the numbers here for this single case will fix this single issue on this specific object, but this might mistakenly break the rendering of otherwise perfectly working objects here and there.

We cannot just pseudo-randomly stab numbers via trial-and-error there and over-fit the loader's behavior around single specific cases and just hope everything else goes well for the rest of the set, we need to have a better idea of the pattern at work behind the issue here.

![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/13f890cf-194f-4908-8425-3bcf00c340f8)

Reply:
My guess is that we also need to be careful of rotate values (if not all) and not just the scaling ones.

Reply:
Something I want to make sure of: did you try first (via three.js for example) to compute the determinant or try to invert the transform matrices affecting those specific polygons with a faulty dark rendering?

Did you try to isolate those specific part of the geometry by fiddling with a copy the RWX object and see how it behaves if you replaced zero values with non-zero ones?

Because I don't really understand how you've reached this very specific explanation about determinants and invertible matrices, especially since the values you are tuning are obviously not purely scaling ones.

Do you need to tune those very specific entries in the `tprops` array to get this result or does it work if you do just `[1, 9]` for example?

Reply:
> Something I want to make sure of: did you try first (via three.js for example) to compute the determinant or try to invert the transform matrices affecting those specific polygons with a faulty dark rendering?

Yes, but the determinant is never exactly 0 after the `fromArray()` call.

> Did you try to isolate those specific part of the geometry by fiddling with a copy the RWX object and see how it behaves if you replaced zero values with non-zero ones?

This is exactly what I did before creating this PR.
 
> Because I don't really understand how you've reached this very specific explanation about determinants and invertible matrices, especially since the values you are tuning are obviously not purely scaling ones.

Since it's a transform matrix, I assumed there were operations on it and I know that for example the invert() function returns a zero matrix if the determinant is 0. My problem being solved by changing zero values to non-zero values, I thought it could be related to something similar.

> Do you need to tune those very specific entries in the `tprops` array to get this result or does it work if you do just `[1, 9]` for example?

`[1, 5]` , `[1, 9]` and `[5, 9]` also work.

Reply:
Ok, then if the core if the issue is that some transformation matrices are not invertible (which sounds really plausible given the quality of some objects...) then maybe a better fix would be to test this first by instantiating a Matrix4 object on the spot and get the determinant value.

If said determinant is nearing 0 (either positively or negatively, but really close to 0), then when can try two things:
- Replace ALL zero members by something really small and hope it becomes invertible then;
- Set this matrix as the identity matrix (`identity()` on `Matrix4`) so it will basically have no effect, I'm not sure how things will look so it we'll need to see, but ideally if this works I'd rather have us do this than modifying values "just in case".

Whatever we pick: at least it will limit the scope of this to matrices that are indeed not invertible to begin with, the rest of the working matrices won't be affected.

Reply:
OR, eventually, what can be tried is to extract the position, scale and rotation information from those faulty matrices (which should give respectively two `Vector3`and one `Euler`) and try to rebuild a new matrix from it, but I have an itch that invertibility may be a required condition for at least one of those operations, still worth a shot though.

See [Extracting position, rotation and scale](https://threejs.org/docs/#api/en/math/Matrix4)

Reply:
Maybe we're not looking at the right part of the code because even if the texture doesn't work, the alphaMap does.
For fun, I reverted my code and tried to comment the line `ctx.rootGroup.applyMatrix4( scale_ten );` instead and this is what I got:
![1691786984 944203041](https://github.com/Blaxar/three-rwx-loader/assets/8944013/e3d3f8dd-e867-4e9c-a462-c52517b2d320)
Which means the texture data might actually be lost during this x10 scaling.

Reply:
Ok, this is somehow reassuring because it's hinting at a race condition or something akin, given that technically the scaling is applied while promises for material update are still likely ongoing.

I still don't specifically see how/why things are overlapping data-wise, but it's less nebulous to comprehend than some foggy matrix determinant issue.

If I move the 10x scaling right after the creation of the root group (instead of doing it at the end of the whole parsing) then for me the issue is gone:

```
diff --git a/src/RWXLoader.js b/src/RWXLoader.js
index e4bfe6b..42a6c12 100644
--- a/src/RWXLoader.js
+++ b/src/RWXLoader.js
@@ -1992,6 +1992,10 @@ class RWXLoader extends Loader {
 
                // Ready root object group
                ctx.rootGroup = new Group();
+
+               // Take the decameter unit into account
+               ctx.rootGroup.applyMatrix4( scale_ten );
+
                ctx.rootGroup.userData.rwx = { axisAlignment: 'none' };
                ctx.currentGroup = ctx.rootGroup;
                ctx.materialStack.push( ctx.materialTracker.currentMaterial );
@@ -2754,9 +2758,6 @@ class RWXLoader extends Loader {
 
                ctx.materialTracker.clearCurrentMaterialList();
 
-               // We're done, return the root group to get the whole object, we take the decameter unit into account
-               ctx.rootGroup.applyMatrix4( scale_ten );
-
                if ( this.waitFullLoad ) {
 
                        // Wait all mask futures before returning loaded object
```

Reply:
Did you test it? Because it doesn't change a thing in Lemuria and I don't think I'm doing anything special on the materials for these cases.
As before, only replacing a value avoids this issue, for example:

```
scale_ten.makeScale( 10.0001, 10, 10 );
```

Reply:
I did, I've compared the rendering before/after the change in a scrap three.js scene but now I can't get the working behavior back, it was just a fluke then...

The worst is: it's not even clear which values give something ok and which doesn't, for example this is `makeScale(2.55, 2.55, 2.55)`
![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/2feff40f-c43a-438d-b99c-491c829679ae)

And this is `makeScale(2.5, 2.5, 2.5)`:
![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/b0881d7f-c726-4e22-800a-b7de6e916f88)

Reply:
Also I've been fiddling with this lovely `arbre12.rwx` to check for myself what the deal with determinant value was:

- I compute the determinant at the end of each `transform` statement
- If the determinant is below `1`: I ignore the data and don't set the matrix, otherwise I take the data in, this is what I get:

```
| Determinant is 0.9515413054205761
|--> Matrix is NOT invertible
***
| Determinant is 30.375
|--> Matrix is invertible
***
| Determinant is 30.375
|--> Matrix is invertible
***
| Determinant is 30.374999707141498
|--> Matrix is invertible
***
| Determinant is 30.3749915593515
|--> Matrix is invertible
***
| Determinant is 30.374991671229
|--> Matrix is invertible
***
| Determinant is 18.37499475 
|--> Matrix is invertible
***
| Determinant is 23.999998376737498
|--> Matrix is invertible
***
| Determinant is 8.999997074606002
|--> Matrix is invertible
***
| Determinant is 7.499997742541
|--> Matrix is invertible
***
While it's true there is one determinant lower than the rest (the very first one actually), if we just ignore this matrix and keep rendering, this is what we get:
![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/1d27e668-6b21-4ca2-87a6-23d9c69a462f)

It just disables part of the trunk, because this transformation matrix likely targets it, but the geometry with the faulty rendering on the texture is still there (because its determinant passed the check), so are you sure you did isolate this in this object step by step?

The is the content of the `tprops` array for which we get a low determinant.

```
Array(16) [ 0.402, 0, 0, 0, 0, 0.401836, 0, 0, 0, 0, … ]
0: 0.402
1: 0
2: 0
3: 0
4: 0
5: 0.401836
6: 0
7: 0
8: 0
9: 0
10: 5.890508
11: 0
12: 0
13: 0
14: 5.890508
15: 1
length: 16
<prototype>: Array []
```

Looking at the file: this is where the transform matching this takes place (see `Transform 0.402 ...`):

```
ProtoBegin cylinder_4_44
  Surface 1. 1. 0.
  TextureModes Lit Foreshorten
  Transform 0.402 0. 0. 0. 0. 0.401836 0. 0. 0. 0. 5.890508 0. 0. 0. 5.890508 1.
  Vertex 0.973042 0. -0.898141 UV 1. 2.847212 # 1
  Vertex 0.19669 0. 0.999996 UV 1. 0. # 2
  Vertex 0.098345 0.170338 0.999996 UV 1.166638 0. # 3
  Vertex 0.489308 0.847504 -0.898141 UV 1.166638 2.847212 # 4
  Vertex -0.098345 0.170338 0.999996 UV 1.333362 0. # 5
  Vertex -0.489262 0.847425 -0.898141 UV 1.333362 2.847212 # 6
  Vertex -0.19669 0. 0.999996 UV 1.5 0. # 7
  Vertex -0.978574 0. -0.898141 UV 1.5 2.847212 # 8
  Vertex -0.098345 -0.170338 0.999996 UV 1.666638 0. # 9
  Vertex -0.488395 -0.845924 -0.898141 UV 1.666638 2.847212 # 10
  Vertex 0.098345 -0.170338 0.999996 UV 1.833362 0. # 11
  Vertex 0.486291 -0.842279 -0.898141 UV 1.833362 2.847212 # 12
  Vertex 0.486291 -0.842279 -0.898141 UV 0.833362 2.847212 # 13
  Vertex 0.098345 -0.170338 0.999996 UV 0.833362 0. # 14
  Vertex -0.19669 0. 0.999996 UV 0.5 0. # 15
  Vertex -0.098345 0.170338 0.999996 UV 0.333362 0. # 16
  Vertex 0.098345 0.170338 0.999996 UV 0.166638 0. # 17
  Vertex -0.098345 -0.170338 0.999996 UV 0.666638 0. # 18
  Vertex -0.678624 1.175406 -1. UV 1.333362 3. # 19
  Vertex -1.357248 0. -1. UV 1.5 3. # 20
  Vertex -0.678624 -1.175406 -1. UV 1.666638 3. # 21
  Vertex 0.678624 -1.175406 -1. UV 1.833362 3. # 22
  Vertex 1.357248 0. -1. UV 1. 3. # 23
  Vertex 0.678624 1.175406 -1. UV 1.166638 3. # 24
  Vertex 0.678624 -1.175406 -1. UV 0.833362 3. # 25
  Color 1. 1. 1.
  Opacity 1.
  Texture maplebrk
  LightSampling Vertex
  Quad 1 23 24 4
  Quad 13 25 23 1
  Quad 10 21 22 12
  Quad 10 8 20 21
  Quad 8 6 19 20
  Quad 6 4 24 19
  Quad 1 2 14 13
  Quad 12 11 9 10
  Quad 8 10 9 7
  Quad 6 8 7 5
  Quad 4 6 5 3
  Quad 4 3 2 1
 ProtoEnd # cylinder_4_44
```

Which, as already expected from the rendering, concerns only the trunk with the maple bark texture (`Texture maplebrk`) but has nothing to do with the buggy foliage (hence probably nothing to do with the determinant value itself).

Reply:
Also I've been fiddling with this lovely `arbre12.rwx` to check for myself what the deal with determinant value was:

- I compute the determinant at the end of each `transform` statement
- If the determinant is below `1`: I ignore the data and don't set the matrix, otherwise I take the data in, this is what I get:

```
| Determinant is 0.9515413054205761
|--> Matrix is NOT invertible
***
| Determinant is 30.375
|--> Matrix is invertible
***
| Determinant is 30.375
|--> Matrix is invertible
***
| Determinant is 30.374999707141498
|--> Matrix is invertible
***
| Determinant is 30.3749915593515
|--> Matrix is invertible
***
| Determinant is 30.374991671229
|--> Matrix is invertible
***
| Determinant is 18.37499475 
|--> Matrix is invertible
***
| Determinant is 23.999998376737498
|--> Matrix is invertible
***
| Determinant is 8.999997074606002
|--> Matrix is invertible
***
| Determinant is 7.499997742541
|--> Matrix is invertible
***
```

While it's true there is one determinant lower than the rest (the very first one actually), if we just ignore this matrix and keep rendering, this is what we get:
![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/1d27e668-6b21-4ca2-87a6-23d9c69a462f)

It just disables part of the trunk, because this transformation matrix likely targets it, but the geometry with the faulty rendering on the texture is still there (because its determinant passed the check), so are you sure you did isolate this in this object step by step?

The is the content of the `tprops` array for which we get a low determinant.

```
Array(16) [ 0.402, 0, 0, 0, 0, 0.401836, 0, 0, 0, 0, … ]
0: 0.402
1: 0
2: 0
3: 0
4: 0
5: 0.401836
6: 0
7: 0
8: 0
9: 0
10: 5.890508
11: 0
12: 0
13: 0
14: 5.890508
15: 1
length: 16
<prototype>: Array []
```

Looking at the file: this is where the transform matching this takes place (see `Transform 0.402 ...`):

```
ProtoBegin cylinder_4_44
  Surface 1. 1. 0.
  TextureModes Lit Foreshorten
  Transform 0.402 0. 0. 0. 0. 0.401836 0. 0. 0. 0. 5.890508 0. 0. 0. 5.890508 1.
  Vertex 0.973042 0. -0.898141 UV 1. 2.847212 # 1
  Vertex 0.19669 0. 0.999996 UV 1. 0. # 2
  Vertex 0.098345 0.170338 0.999996 UV 1.166638 0. # 3
  Vertex 0.489308 0.847504 -0.898141 UV 1.166638 2.847212 # 4
  Vertex -0.098345 0.170338 0.999996 UV 1.333362 0. # 5
  Vertex -0.489262 0.847425 -0.898141 UV 1.333362 2.847212 # 6
  Vertex -0.19669 0. 0.999996 UV 1.5 0. # 7
  Vertex -0.978574 0. -0.898141 UV 1.5 2.847212 # 8
  Vertex -0.098345 -0.170338 0.999996 UV 1.666638 0. # 9
  Vertex -0.488395 -0.845924 -0.898141 UV 1.666638 2.847212 # 10
  Vertex 0.098345 -0.170338 0.999996 UV 1.833362 0. # 11
  Vertex 0.486291 -0.842279 -0.898141 UV 1.833362 2.847212 # 12
  Vertex 0.486291 -0.842279 -0.898141 UV 0.833362 2.847212 # 13
  Vertex 0.098345 -0.170338 0.999996 UV 0.833362 0. # 14
  Vertex -0.19669 0. 0.999996 UV 0.5 0. # 15
  Vertex -0.098345 0.170338 0.999996 UV 0.333362 0. # 16
  Vertex 0.098345 0.170338 0.999996 UV 0.166638 0. # 17
  Vertex -0.098345 -0.170338 0.999996 UV 0.666638 0. # 18
  Vertex -0.678624 1.175406 -1. UV 1.333362 3. # 19
  Vertex -1.357248 0. -1. UV 1.5 3. # 20
  Vertex -0.678624 -1.175406 -1. UV 1.666638 3. # 21
  Vertex 0.678624 -1.175406 -1. UV 1.833362 3. # 22
  Vertex 1.357248 0. -1. UV 1. 3. # 23
  Vertex 0.678624 1.175406 -1. UV 1.166638 3. # 24
  Vertex 0.678624 -1.175406 -1. UV 0.833362 3. # 25
  Color 1. 1. 1.
  Opacity 1.
  Texture maplebrk
  LightSampling Vertex
  Quad 1 23 24 4
  Quad 13 25 23 1
  Quad 10 21 22 12
  Quad 10 8 20 21
  Quad 8 6 19 20
  Quad 6 4 24 19
  Quad 1 2 14 13
  Quad 12 11 9 10
  Quad 8 10 9 7
  Quad 6 8 7 5
  Quad 4 6 5 3
  Quad 4 3 2 1
 ProtoEnd # cylinder_4_44
```

Which, as already expected from the rendering, concerns only the trunk with the maple bark texture (`Texture maplebrk`) but has nothing to do with the buggy foliage (hence probably nothing to do with the determinant value itself).

Reply:
Update: forcing a `Facet` light sampling for those specific foliage polygons does the trick.

```
ProtoBegin plane_37_45
  Surface 1. 1. 0.
  TextureModes Lit Foreshorten
  Transform 0. -4.5 0. 0. 4.5 0. 0. 0. 0. 0. 1.5 0. 0. 0. 6.75 1.
  Vertex 0. 0. 0. UV 0.5 0.5 # 1
  Vertex 1. -1. 0. UV 0. 1. # 2
  Vertex -1. -1. -1.5 UV 0. 0. # 3
  Vertex -1. 1. 0. UV 1. 0. # 4
  Vertex 1. 1. -1.5 UV 1. 1. # 5
  Color 1. 1. 1.
  Opacity 1.
  Texture maple4a Mask maple4m
  LightSampling Facet
  Quad 1 4 5 2
  Quad 2 5 4 1
  Quad 1 2 3 4
  Quad 4 3 2 1
 ProtoEnd # plane_37_45
 ProtoBegin plane_44_46
  Surface 1. 1. 0.
  TextureModes Lit Foreshorten
  Transform 4.5 0. 0. 0. 0. 4.5 0. 0. 0. 0. 1.5 0. 0. 0. 6. 1.
  Vertex 0. 0. 0. UV 0.5 0.5 # 1
  Vertex 1. -1. 0. UV 0. 1. # 2
  Vertex -1. -1. -1.5 UV 0. 0. # 3
  Vertex -1. 1. 0. UV 1. 0. # 4
  Vertex 1. 1. -1.5 UV 1. 1. # 5
  Color 1. 1. 1.
  Opacity 1.
  Texture maple4a Mask maple4m
  LightSampling Facet
  Quad 1 4 5 2
  Quad 2 5 4 1
  Quad 1 2 3 4
  Quad 4 3 2 1
 ProtoEnd # plane_44_46
```

![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/fcba58d0-ed76-4e22-b259-02a968389809)

This confirms my itch that it had something to do with normals ans light behavior, but it's still not clear what...
This probes more questions about the value of the vertex normals themselves.

If we isolate the faulty geometry, we get this:

![image](https://github.com/Blaxar/three-rwx-loader/assets/5609999/64837576-c68a-4577-b2c4-cae28bf5ec67)

If we look at the normals, they do look off:

```
Object { position: {…}, uv: {…}, normal: {…} }
​normal: Object { isBufferAttribute: true, itemSize: 3, count: 10, … }
​​array: Float32Array(30) [ 0, -1, 0, … ]
​​​0: 0
​​​1: -1
​​​2: 0
​​​3: 0
​​​4: 1
​​​5: 0
​​​6: 0
​​​7: 0
​​​8: 0
​​​9: 0
​​​10: 1
​​​11: 0
​​​12: 0
​​​13: 0
​​​14: 0
​​​15: 0
​​​16: -1
​​​17: 0
​​​18: 0
​​​19: 1
​​​20: 0
​​​21: 0
​​​22: 0
​​​23: 0
​​​24: 0
​​​25: 1
​​​26: 0
​​​27: 0
​​​28: 0
​​​29: 0
​​​buffer: ArrayBuffer { byteLength: 120 }
​​​byteLength: 120
​​​byteOffset: 0
​​​length: 30
​​​<prototype>: Float32Array.prototype { … }
​​count: 10
​​isBufferAttribute: true
​​itemSize: 3
name: ""
normalized: false
```

Reply:
A quick update before I forget: the solution doesn't involve tweaking the transformation matrices at all, but rather detecting when normal values are off (0-length vectors) when declaring faces in quads (and possibly triangles), so we can duplicate vertices on the spot to use them for the newly-defined faces instead of reusing the existing vertices from the current set.

What's happening for this object is: when there are two quads using the same set of vertices, one to point up and the other one to point down (giving two vectors of exact opposite directions) then they nullify each other when computing the averaged normal for those vertices (`.computeVertexNormals()`).

You can test this by disabling some `Quad` statements from the dark foliage and see that, as one side of the foliage will just disappear, then the other will become properly textured and lit up since the normals are turning ok. 

The fact that some transformation matrices gave better result is just sheer luck on the inner computation, dealing with an otherwise not-so-well-formed 3D geometry (with self-contradicting requirements), note also that if we disable flattening (`.setFlatten( false )` on the loader) then even tweaking the root scaling value won't work anymore.

Given that the proper solution and its sturdy implementation (more complicated, not going to lie) have nothing to do with the content of this PR: do you still intend to work on it and update things to match the needs or can we close it?

Reply:
I guess I'll just close it then, checking quads with opposing normals and split the vertices is indeed a big change that involves face indices updates. Considering it only affects some models, it might not be worth it in the end.